File: finger_exercises/chapter4/find_root_sum.py
"""Use the find root function to print the sum of approximations
to the square root of 25, cube root of -8 and fourth root of 16,
use epsilon 0.001"""

def find_root(x, power, epsilon):
    """find interval containing answer"""
    if x < 0 and (power % 2 == 0):
        return None #negative number has no even-powered root
    low = min(-1,x)
    high = max(1, x)
    # Use bisection search
    ans = (high + low) / 2
    while (abs(ans**power - x) >= epsilon):
        if ans ** power < x:
            low = ans
        else:
            high = ans
        ans = (high + low) / 2
    return ans

e = 0.001
print(sum([find_root(25, 2, e), find_root(-8, 3, e), find_root(16, 4, e)]))

File: finger_exercises/chapter4/find_root_sum.c
#include <stdio.h>
#include <math.h>

float find_root(int x, int power, float epsilon)
{
    float low;
    float high;
    float ans;
    if (x < -1)
    {
        low = x;
    }
    else
    {
        low = -1;
    }
    if (x > 1)
    {
        high = x;
    }
    else
    {
        high = 1;
    }
    ans = (high + low) / 2;
    //negative numbers do not have an even-powered root
    if ((x < 0) && (power % 2 == 0))
    {
        printf("Not possible\n");
        return 0;
    }
    while (fabs(pow(ans, power) - x) > epsilon)
    {
        if (pow(ans, power) > x)
        {
            high = ans;
        }
        else
        {
            low = ans;
        }
        ans = (high + low) / 2;
    }
    return ans;


}

int main(void)
{
    int xs[] = {25, -8, 16};
    int powers[] = {2,3,4};
    float epsilon = 0.001;
    float ans;
    float sum;
    for (int i = 0; i < 3; i++)
    {
        ans = find_root(xs[i], powers[i], epsilon);
        if (ans != 0)
        {
            sum += ans;
        }
    }
    printf("%f\n", sum);
    return 0;
}
